cs-quicklabs/quick-learn
# Learning dashboard Ui and api

# Add User Roadmaps Feature

## What does this PR do?
This PR implements a new feature that allows users to view their assigned roadmaps and associated courses on the dashboard. It includes both backend API implementation and frontend integration.

### Backend Changes
- Added new `GET /api/v1/users/my-roadmaps` endpoint to fetch user's assigned roadmaps
- Implemented roadmap fetch with optional course inclusion using query parameter
- Added progress indicators (currently hardcoded to 100% for MVP)
- Proper route ordering to avoid UUID parsing conflicts
- Maintained consistent error handling with existing patterns

### Frontend Changes
- Updated Dashboard component to fetch and display real roadmap data
- Added types for roadmap and course responses
- Integrated with existing UI components
- Added loading and error states
- Updated profile service to include roadmap fetching
- Implemented ghost loader for better UX during data fetching
- Added responsive grid layout (2-5 columns based on screen size)
- Routes configured: 
  - Roadmaps: `/dashboard/content/:roadmapId`
  - Courses: `/dashboard/content/courses/:courseId`

### UI/UX Improvements
- Added skeleton loading state matching the final UI layout
- Implemented hover effects on cards for better interactivity
- Added progress bars with percentage indicators
- Truncated long titles and descriptions with ellipsis
- Maintained consistent spacing and layout with design system

### API Details
```typescript
GET /api/v1/users/my-roadmaps?include_courses=true
Response:
{
  "success": true,
  "message": "Successfully retrieved user roadmaps",
  "data": [
    {
      "id": number,
      "name": string,
      "description": string,
      "percentage": number,
      "courses": [
        {
          "id": number,
          "name": string,
          "description": string,
          "percentage": number
        }
      ]
    }
  ]
}
```

### Types Added
```typescript
export type TUserCourse = {
  id: number;
  name: string;
  description: string;
  percentage?: number;
  lessons?: TUserLesson[];
};

export type TUserLesson = {
  id: number;
  name: string;
  content: string;
  completed: boolean;
};
```

## How should this be tested?
1. Login to the application
2. Navigate to the dashboard
3. Verify that assigned roadmaps are displayed
4. Verify that courses within roadmaps are displayed correctly
5. Check that UI properly handles:
   - Loading states with skeleton loaders
   - Empty states
   - Error states
6. Verify responsive behavior:
   - Mobile (2 columns)
   - Tablet (3 columns)
   - Desktop (4 columns)
   - Large Desktop (5 columns)
7. Test navigation:
   - Click roadmap card -> should navigate to `/dashboard/content/:roadmapId`
   - Click course card -> should navigate to `/dashboard/content/courses/:courseId`

## Future Improvements
- [ ] Implement actual progress calculation
- [ ] Add pagination for roadmaps and courses
- [ ] Add filtering and sorting options
- [ ] Add caching for better performance
- [ ] Add progress history tracking
- [ ] Add search functionality for roadmaps and courses
- [ ] Implement drag and drop for reordering
- [ ] Add course completion tracking
- [ ] Implement roadmap sharing functionality
- [ ] Add course recommendations based on progress

## Dependencies
- No new dependencies added
- Uses existing UI components and services
- Integrates with existing translation system

## Notes
- Progress percentages are currently hardcoded to 100% as placeholder
- This is the first phase of roadmap tracking feature
- UI is consistent with existing design patterns
- All translations are managed through the en.ts file
- Maintained existing error handling patterns
- Used Tailwind CSS for styling consistency
- Implemented proper TypeScript types throughout

## Screenshots
![image](https://github.com/user-attachments/assets/2c32b45e-cdfd-40ea-83d5-8269e1b72a18)
![image](https://github.com/user-attachments/assets/3b422fbd-afcf-45ac-bcfa-c87ac968e16f)
